governator-test-junit/src/main/java/com/netflix/governator/guice/test/AnnotationBasedTestInjectorManager.java

```diff
@@ -0,0 +1,223 @@
+
+        moduleClasses.addAll(getModulesForAnnotatedClass(testClass));
+        for (Class<?> parentClass : getAllSuperClassesInReverseOrder(testClass)) {
+            moduleClasses.addAll(getModulesForAnnotatedClass(parentClass));
+        }
+        for (Class<? extends Module> moduleClass : moduleClasses) {
+            try {
+                modulesForTestClass.add(moduleClass.newInstance());
+            } catch (InstantiationException | IllegalAccessException e) {
+                try {
+                    Constructor<?> constructor = moduleClass.getDeclaredConstructors()[0];
+                    constructor.setAccessible(true);
+                    modulesForTestClass.add((Module) constructor.newInstance());
+                } catch (Exception ex) {
+                    throw new RuntimeException("Error instantiating module " + moduleClass
+                            + ". Please ensure that the module is public and has a no-arg constructor", e);
+                }
+            }
+        }
+    }
+
+    private List<Class<? extends Module>> getModulesForAnnotatedClass(Class<?> testClass) {
+        final Annotation annotation = testClass.getAnnotation(ModulesForTesting.class);
+        if (annotation != null) {
+            return Arrays.asList(((ModulesForTesting) annotation).value());
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    private void inspectMocksForTestClass(Class<?> testClass) {
+        getMocksForAnnotatedFields(testClass);
+        for (Class<?> parentClass : getAllSuperClassesInReverseOrder(testClass)) {
+            getMocksForAnnotatedFields(parentClass);
+        }
+    }
+
+    private void getMocksForAnnotatedFields(Class<?> testClass) {
+
+        for (Field field : testClass.getDeclaredFields()) {
+            if (field.isAnnotationPresent(ReplaceWithMock.class)) {
+                overrideModules.add(new MockitoOverrideModule<>(field.getAnnotation(ReplaceWithMock.class), field.getType()));
+            }
+            if (field.isAnnotationPresent(WrapWithSpy.class)) {
+                WrapWithSpy spyAnnotation = field.getAnnotation(WrapWithSpy.class);
+                if (spyAnnotation.name().isEmpty()) {
+                    spyTargets.add(Key.get(field.getType()));
+                }
+                else {
+                    spyTargets.add(Key.get(field.getType(), Names.named(spyAnnotation.name())));
+                }
+            }
+        }
+    }
+
+    private void inspectSpiesForTargetKeys(List<Element> elements) {
+        for (Element element : elements) {
+            element.acceptVisitor(new DefaultElementVisitor<Void>() {
+                @Override
+                public <T> Void visit(Binding<T> binding) {
+                    final Binding<T> finalBinding = binding;
+                    if (spyTargets.contains(binding.getKey())) {
+                        AbstractModule spyModule = new AbstractModule() {
```

Comment:
Consider lifting this anonymous module out and naming it, akin to MockitoOverrideModule.  This block is moderately convoluted.